### PR TITLE
Current mounting code

### PR DESCRIPTION
**Pull request by DoctorMacc**
- Bedrock players visually see mounting
- Boat rowing animation works on both sides
- Minecarts are 100% done, I believe.

TODO:

- Proper boat movement from Bedrock to Java
- Horses do not operate and kick you for flying
- Register Bedrock EntityLinks on chunk load and login.

### CODE REVIEW

File: connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java [hunk 1]
```diff
@@ -26,28 +26,107 @@
...
 import org.geysermc.connector.network.session.GeyserSession;
 
+import java.util.concurrent.TimeUnit;
+
 public class BoatEntity extends Entity {
 
+    private boolean is_paddling_left;
+    private float paddle_time_left;
+    private boolean is_paddling_right;
+    private float paddle_time_right;
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java:43]: Variable name formatting should be in `camelCase`, not `snake_case`.

File: connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java [hunk 2]
```diff
@@ -26,28 +26,107 @@
...
         super.updateBedrockMetadata(entityMetadata, session);
     }
+
+    public void updateLeftPaddle(GeyserSession session, EntityMetadata entityMetadata) {
+        while (is_paddling_left) {
+            try {
+                paddle_time_left += ROWING_SPEED;
+                metadata.put(EntityData.PADDLE_TIME_LEFT, paddle_time_left);
+                super.updateBedrockMetadata(entityMetadata, session);
+                Thread.sleep(100);
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java:105]: Should not be sleeping the thread at all here as this sleeps every other operation too.

File: connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java [hunk 3]
```diff
@@ -26,28 +26,107 @@
...
+            try {
+                paddle_time_right += ROWING_SPEED;
+                metadata.put(EntityData.PADDLE_TIME_RIGHT, paddle_time_right);
+                super.updateBedrockMetadata(entityMetadata, session);
+                // Something I tried recently, doesn't really work - keeping just in case someone wants to try it.
+//                ClientVehicleMovePacket clientVehicleMovePacket = new ClientVehicleMovePacket(
+//                        entity.position.getX(), entity.position.getY(), entity.position.getZ(), entity.getBedrockRotation().getX() + 10f, 0
+//                );
+//                session.getDownstream().getSession().send(clientVehicleMovePacket);
+                Thread.sleep(100);
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java:125]: Same as above.

File: connector/src/main/java/org/geysermc/connector/entity/Entity.java [hunk 4]
```diff
@@ -112,6 +109,14 @@ public void spawnEntity(GeyserSession session) {
         addEntityPacket.setRotation(getBedrockRotation());
         addEntityPacket.setEntityType(entityType.getType());
         addEntityPacket.getMetadata().putAll(metadata);
+        //USELESS CODE START
+        Integer[] passengerIds = session.getEntityCache().getCachedEntityLink((int) entityId);
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/entity/Entity.java:113]: Should be `int[]`.

File: connector/src/main/java/org/geysermc/connector/entity/PlayerEntity.java [hunk 5]
```diff
@@ -85,6 +86,15 @@ public void spawnEntity(GeyserSession session) {
         addPlayerPacket.setPlatformChatId("");
         addPlayerPacket.getMetadata().putAll(metadata);
 
+        //USELESS CODE START
+        Integer[] passengerIds = session.getEntityCache().getCachedEntityLink((int) entityId);
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/entity/PlayerEntity.java:90]: Same as above.

File: connector/src/main/java/org/geysermc/connector/network/session/cache/EntityCache.java [hunk 6]
```diff
@@ -49,6 +50,7 @@
     private Long2LongMap entityIdTranslations = Long2LongMaps.synchronize(new Long2LongOpenHashMap());
     private Map<UUID, PlayerEntity> playerEntities = Collections.synchronizedMap(new HashMap<>());
     private Object2LongMap<UUID> bossbars = new Object2LongOpenHashMap<>();
+    private Int2ObjectOpenHashMap<Integer[]> cachedEntityLinks = new Int2ObjectOpenHashMap<>();
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/session/cache/EntityCache.java:53]: Should be `private Int2ObjectMap<int[]> cachedEntityLinks = new Int2ObjectOpenHashMap<>()`.

File: connector/src/main/java/org/geysermc/connector/network/session/cache/EntityCache.java [hunk 7]
```diff
@@ -136,4 +138,12 @@ public void clear() {
         playerEntities = null;
         bossbars = null;
     }
+
+    public Integer[] getCachedEntityLink(int entityId) {
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/session/cache/EntityCache.java:142]: Same as above.

File: connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockAnimateTranslator.java [hunk 8]
```diff
@@ -37,19 +37,26 @@
 @Translator(packet = AnimatePacket.class)
 public class BedrockAnimateTranslator extends PacketTranslator<AnimatePacket> {
 
+    private boolean is_steering_left;
+    private boolean is_steering_right;
```
Comment (RednedEpic) [inline connector/src/main/java/org/geysermc/connector/network/translators/bedrock/BedrockAnimateTranslator.java:41]: Variable name formatting should be in camelCase, not snake_case, and this data should be stored in `GeyserSession`.

**Comment by JuniorJPDJ** [comment]
does it fix also bed usage?

**Comment by DoctorMacc** [comment]
The current mounting code does not anticipate to fix bed usage.

**Pull request merged**